### PR TITLE
Reader Improvements: Adds ability to save interests from the Select Interests view

### PR DESCRIPTION
**Pull request by emilylaguna**
Main Issue: https://github.com/wordpress-mobile/WordPress-iOS/issues/14396

WordPressKit Related PR: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/262

### Description
This PR includes some major changes:
- The select interest view is now displayed over the Discover tab's content
- There is a new "loading" state displayed when loading the interests, and saving them
- After selecting some interests on the select interests view they will be saved:
  - If the user is logged into a WP.com account they will be saved remotely
  - If the user is not logged in (self hosted) the interests will be saved locally
- There is no "mark as seen" state for the select interests view, if the user has no interests added they will be presented with the picker

### If the user is in a self hosted/logged out state
- Interests will be stored locally, but if they later sign into a WP.com account the local interests will be removed and replaced with the ones from the server
- The "recommended" interests no longer appear since it will be replaced by the select interests view on the Discover page
- I removed the 'add' button from the 'Tags' tab in the filter bar since adding of remote tags is not allowed when the user is self hosted

### Screenshots
| Select Interests | Loading |
|---|---|
|||

### To test:

**Before testing you should delete and resintall the app to reset the "seen" user default**

#### 01 - Select Interests Picker
1. Login using a WP.com account
2. Launch the app
3. Delete all of your followed tags if you have any
4. Tap on the Reader tab
5. Tap on the 'Discover' tab

**Expectation:** You are presented with the Select Interests View

#### 02 - Save Interests: Loading View
1. Continuing with the steps from 01
2. Select some interests
3. Tap the 'Done' button

**Expectation:** The select interests view fades out, and you're presented with a loading view with an animated spinner and a message

#### 03 - Save Interests
1. Continuing with the steps from 02
2. Once the request completes, the loading view will fade out
3. Tap on the Following tab
4. Tap on the 'Cog'
5. Tap on the 'Followed Tags' 

**Expectation:** You should see all the interests your selected listed in the table
**Bonus Test:** Go to https://wordpress.com/read you should see all the interests listed in the sidebar as well

#### 04 - You don't see the picker again
1. Continuing with the steps from 03
2. Tap back to the Discover tab

**Expectation:** You see the discover feed

#### 04 - Removing all interests
1. Continuing with the steps from 04
2. Tap back to the Following tab
3. Remove all of your interests using the 'Cog'
4. Tap back to the 'Discover' tab

**Expectation:** You see the Select Interests view again

#### 01 - You have no followed sites or tags
1. Login using a self hosted site
2. Launch the app
4. Tap on the Reader tab
5. Tap on the 'Follow' tab
6. Tap on the 'Filter' item
7. Tap on the 'Tags' tab

**Expectation:** The title of the tab should be Tags (0), and there should be a message informing you, you have no tags and should add one. 
**Expectation 2:** You should not see an add button

#### 02 - Select Interests Displayed
1. Launch the app with a self hosted site added
2. Tap on the Reader tab
3. Tap on Discover

**Expectation:** You are presented with the select interests view

#### 03 - Select Interests Saved
1. Continuing with the steps from 02
2. Select some interests 
3. Tap the 'Done' button
4. The view should be hidden
5. Tap on the 'Following' tab
6. Tap on the 'Filter' option
7. Tap on the 'Tags' tab

**Expectation:** You should see the list of interests you selected in the table view

#### 04 - Select Interests Work
1. Continuing with the steps from 03
2. Tap on an interest

**Expectation 1:**  The filter should be set to the interest you selected
**Expectation 2:** Posts for that interest should be displayewd

#### 04 - You don't see the picker again
1. Continuing with the steps from 04
2. Tap back to the Discover tab

**Expectation:** You see the discover feed

### PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 1]
```diff
@@ -21,10 +21,12 @@ import WordPressFlux
     @objc static let restorationClassIdentifier = "ReaderStreamViewControllerRestorationIdentifier"
     @objc static let restorableTopicPathKey: String = "RestorableTopicPathKey"
 
-    private var interestsCoordinator = ReaderSelectInterestsCoordinator()
-
     // MARK: - Properties
 
+    // Select Interests
+    private var interestsCoordinator = ReaderSelectInterestsCoordinator()
```
Comment (leandroalonso) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:27]:
```suggestion
    private lazy var interestsCoordinator: ReaderSelectInterestsCoordinator = {
        ReaderSelectInterestsCoordinator()
    }
```

File: WordPress/Classes/Services/ReaderTopicService+FollowedInterests.swift [hunk 2]
```diff
@@ -22,16 +30,69 @@ extension ReaderTopicService: ReaderFollowedInterestsService {
...
+                         success: @escaping ([ReaderTagTopic]?) -> Void,
+                         failure: @escaping (Error) -> Void,
+                         isLoggedIn: Bool) {
+        // Check if the user is logged in
+        guard isLoggedIn else {
+            followInterestsLocally(interests, success: success, failure: failure)
+            return
         }
+
+        //If they're not, attempt to save the interests on the server
```
[the remark this answers is not in the document]
Comment (emilylaguna) [reply WordPress/Classes/Services/ReaderTopicService+FollowedInterests.swift:49]: Ah, this was meant to be read in addition to the comment above:

"Check if the user is logged in, if they're not..." I'll change this.

File: WordPress/Classes/Services/ReaderTopicService+FollowedInterests.swift [hunk 3]
```diff
@@ -22,16 +30,69 @@ extension ReaderTopicService: ReaderFollowedInterestsService {
...
+        return WordPressComRestApi.defaultApi(oAuthToken: token,
+                                              userAgent: WPUserAgent.wordPress())
     }
 
-    // MARK: - Private: Helpers
+    // MARK: - Private: Fetching Helpers
     private func followedInterestsFetchRequest() -> NSFetchRequest<ReaderTagTopic> {
         let entityName = "ReaderTagTopic"
-        let predicate = NSPredicate(format: "following = YES")
+        let predicate = NSPredicate(format: "following = YES AND showInMenu = YES")
```
[the remark this answers is not in the document]
Comment (emilylaguna) [reply WordPress/Classes/Services/ReaderTopicService+FollowedInterests.swift:95]: It's reusing the logic used elsewhere to get the following tags.

File: WordPress/Classes/Services/ReaderTopicService+FollowedInterests.swift [hunk 4]
```diff
@@ -22,16 +30,69 @@ extension ReaderTopicService: ReaderFollowedInterestsService {
...
+
+        interests.forEach { interest in
+            let topic = NSEntityDescription.insertNewObject(forEntityName: "ReaderTagTopic", into: managedObjectContext) as! ReaderTagTopic
+            topic.tagID = ReaderTagTopic.loggedOutTagID
+            topic.type = ReaderTagTopic.TopicType
+            topic.path = service.pathForTopic(slug: interest.slug)
+            topic.following = true
+            topic.showInMenu = true
+            topic.title = interest.title
+            topic.slug = interest.slug
```
[the remark this answers is not in the document]
Comment (emilylaguna) [reply WordPress/Classes/Services/ReaderTopicService+FollowedInterests.swift]: Done in: https://github.com/wordpress-mobile/WordPress-iOS/pull/14487/commits/30ffc119c7e25b93179d6a822eb56ec2c096f90b

File: WordPress/Classes/Services/ReaderTopicService+FollowedInterests.swift [hunk 5]
```diff
@@ -22,16 +30,69 @@ extension ReaderTopicService: ReaderFollowedInterestsService {
...
+        }
+
+        ContextManager.sharedInstance().save(managedObjectContext, withCompletionBlock: { [weak self] in
```
[the remark this answers is not in the document]
Comment (emilylaguna) [reply WordPress/Classes/Services/ReaderTopicService+FollowedInterests.swift:75]: I can't inject the context in this class because it's an extension.

**Comment by leandroalonso** [comment]
@yaelirub could you please review this one too?

File: Podfile.lock [hunk 6]
```diff
@@ -397,7 +397,7 @@ PODS:
     - WordPressKit (~> 4.0-beta.0)
     - WordPressShared (~> 1.9-beta)
     - WordPressUI (~> 1.7.0)
-  - WordPressKit (4.12.1-beta.1):
+  - WordPressKit (4.12.1-beta.2):
```
Comment (yaelirub) [inline Podfile.lock:400]: indentation

File: WordPress/Classes/Services/ReaderTopicService+FollowedInterests.swift [hunk 7]
```diff
@@ -22,16 +30,66 @@ extension ReaderTopicService: ReaderFollowedInterestsService {
...
+            self?.fetchFollowedInterestsLocally(completion: completion)
+        }
+    }
+
+    func followInterests(_ interests: [RemoteReaderInterest],
+                         success: @escaping ([ReaderTagTopic]?) -> Void,
+                         failure: @escaping (Error) -> Void,
+                         isLoggedIn: Bool) {
+        // If the user is not logged in, save the interests locally
+        guard isLoggedIn else {
```
Comment (yaelirub) [inline WordPress/Classes/Services/ReaderTopicService+FollowedInterests.swift:44]: I much prefer an if...else here

File: WordPress/Classes/Services/ReaderTopicService.m [hunk 8]
```diff
@@ -961,6 +961,19 @@ - (void)mergeMenuTopics:(NSArray *)topics withSuccess:(void (^)(void))success
                         // removing the topic, if it was once followed its not now.
                         topic.following = NO;
                     } else {
+                        // If the user adds a locally saved tag/interest prevent it from being deleted
+                        // while the user is logged out.
+                        if([Feature enabled:FeatureFlagReaderImprovementsPhase2]){
```
Comment (yaelirub) [inline WordPress/Classes/Services/ReaderTopicService.m:966]: missing spaces all over

File: WordPress/Classes/Services/ReaderTopicService.m [hunk 9]
```diff
@@ -961,6 +961,19 @@ - (void)mergeMenuTopics:(NSArray *)topics withSuccess:(void (^)(void))success
                         // removing the topic, if it was once followed its not now.
```
Comment (yaelirub) [inline WordPress/Classes/Services/ReaderTopicService.m:961]: I think the comments are unnecessary here

File: WordPress/Classes/Services/ReaderTopicService.m [hunk 10]
```diff
@@ -961,6 +961,19 @@ - (void)mergeMenuTopics:(NSArray *)topics withSuccess:(void (^)(void))success
...
                         topic.following = NO;
                     } else {
+                        // If the user adds a locally saved tag/interest prevent it from being deleted
+                        // while the user is logged out.
+                        if([Feature enabled:FeatureFlagReaderImprovementsPhase2]){
+                            if(!ReaderHelpers.isLoggedIn && [topic isKindOfClass:ReaderTagTopic.class]){
```
Comment (yaelirub) [inline WordPress/Classes/Services/ReaderTopicService.m:967]: Not sure I understand. Why do we need to check if not logged in while we're deleting? Isn't it enough to check that `!tagTopic.wasAddedWhileLoggedOut` before deleting?

File: WordPress/Classes/ViewRelated/Reader/ReaderInterestsCoordinator.swift [hunk 11]
```diff
@@ -34,6 +34,19 @@ class ReaderSelectInterestsCoordinator {
         }()
     }
 
+    // MARK: - Saving
+    public func saveInterests(interests: [RemoteReaderInterest], completion: @escaping (Bool) -> Void) {
+        let isLoggedIn = userId != nil ? true : false
```
Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/ReaderInterestsCoordinator.swift:39]: can remove the `? true : false`

Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/ReaderInterestsCoordinator.swift:39]: Also, why were we using `ReaderHelpers.isLoggedIn()` till now but now we're checking it like this?

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 12]
```diff
@@ -1882,6 +1874,66 @@ extension ReaderStreamViewController: ReaderPostUndoCellDelegate {
 }
 
 
+// MARK: - Select Interests Display
+private extension ReaderStreamViewController {
+    func displaySelectInterestsIfNeeded(_ content: ReaderContent) {
+        guard
```
Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:1880]: lets keep the style consistent and do the guard in the same line as well as the else

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 13]
```diff
@@ -1882,6 +1874,66 @@ extension ReaderStreamViewController: ReaderPostUndoCellDelegate {
...
+            FeatureFlag.readerImprovementsPhase2.enabled,
+            content.topicType == .discover
+        else {
+            // Hide the view if we're not on the discover tab, and it exists
```
Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:1884]: We're not hiding it, we're removing it from parent

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 14]
```diff
@@ -1882,6 +1874,66 @@ extension ReaderStreamViewController: ReaderPostUndoCellDelegate {
...
+            self.selectInterestsViewController?.remove()
+            return
+        }
+
+        if self.selectInterestsViewController != nil {
```
Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:1889]: We're guarding in showSelectInterestsViewIfNeeded so why do we need to check if it's no nil here?

Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:1889]: Why is that needed. Can we not remove this and just do:
```interestsCoordinator.shouldDisplay { [unowned self] shouldDisplay in
            if shouldDisplay {
                self.makeSelectInterestsViewControllerIfNeeded()
                self.showSelectInterestsViewIfNeeded()
            }
        }
```
?

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 15]
```diff
@@ -1882,6 +1874,66 @@ extension ReaderStreamViewController: ReaderPostUndoCellDelegate {
...
+        }
+    }
+
+    func showSelectInterestsViewIfNeeded() {
+        guard let controller = self.selectInterestsViewController else {
+            return
+        }
+
+        UIView.animate(withDuration: 0) {
+            controller.view.frame = self.view.bounds
```
Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:1909]: no need for self

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 16]
```diff
@@ -1882,6 +1874,66 @@ extension ReaderStreamViewController: ReaderPostUndoCellDelegate {
...
+            }
+        }
+    }
+
+    func showSelectInterestsViewIfNeeded() {
+        guard let controller = self.selectInterestsViewController else {
+            return
+        }
+
+        UIView.animate(withDuration: 0) {
```
Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:1908]: Why are we doing animation with 0 duration ?

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 17]
```diff
@@ -1882,6 +1874,66 @@ extension ReaderStreamViewController: ReaderPostUndoCellDelegate {
...
+            controller.view.frame = self.view.bounds
+            self.add(controller, asChildOf: self)
+        }
+    }
+
+    func makeSelectInterestsViewControllerIfNeeded() {
```
Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:1914]: I don't understand this function at all :(

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 18]
```diff
@@ -1882,6 +1874,66 @@ extension ReaderStreamViewController: ReaderPostUndoCellDelegate {
...
+        if selectInterestsViewController != nil {
+            return
+        }
+
+        let controller = ReaderSelectInterestsViewController()
+        controller.didSaveInterests = { [unowned self] in
+            guard let controller = self.selectInterestsViewController else {
```
Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:1921]: But wouldn't selectInterestsViewController be empty? Isn't that what we're checking at the beginning?

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 19]
```diff
@@ -1882,6 +1874,66 @@ extension ReaderStreamViewController: ReaderPostUndoCellDelegate {
...
+        }
+
+        UIView.animate(withDuration: 0) {
+            controller.view.frame = self.view.bounds
+            self.add(controller, asChildOf: self)
+        }
+    }
+
+    func makeSelectInterestsViewControllerIfNeeded() {
+        if selectInterestsViewController != nil {
```
Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:1915]: Guard would be better here

**Comment by yaelirub** [comment]
A couple of noticeable issues here:
1. When selecting interests while offline, I get stuck on an empty interests collection view with no option to refresh even when network is back


2. Sometimes the app would open on the Discover tab. If I don't have any interests  it would show the "old" Discover content and only when moving back and forth between the tabs does it show the interests screen

File: WordPress/Classes/Services/ReaderTopicService.m (hunk 10 above)
Comment (leandroalonso) [reply WordPress/Classes/Services/ReaderTopicService.m]: This method is called if the user is logged in or not. If we don't check if the user `isLoggedIn` here, a logged out user will have the followed topics deleted each time.

Basically, what we want to achieve here is to delete followed topics only if the user is now logged in.

File: WordPress/Classes/ViewRelated/Reader/ReaderInterestsCoordinator.swift (hunk 11 above)
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Reader/ReaderInterestsCoordinator.swift]: Both do the same.

I'd say that the advantage here is that since we have the `userId` we can rely directly on this property instead of calling `ReaderHelpers.isLoggedIn()  — which will rely on Core Data again.

Anyway, both will work. Do you prefer to change? Is that confusing to you?

**Comment by leandroalonso** [comment]
@yaelirub thanks for the review!

About 1) I'll fix it.

About 2) can we leave that for a subsequent task? I have a refactoring planned that will probably fix that too. :)

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift (hunk 16 above)
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:1907]: I tested here to discover:

1) With 0 duration: https://d.pr/v/QWiGIG
2) Without: https://d.pr/v/jQGbip

You can see that without 0 duration the screen quickly blinks. I added `withDuration: 0` back plus added a comment to explain why it's needed in 77fd1ff4374fc05b1309e0b08a21bf0479dd4fef

**Comment by leandroalonso** [comment]
@yaelirub ready for another review!

I added two error handlings:

### When the list of interests fails to load



### When saving the interests fails



Just a reminder that this:

> Sometimes the app would open on the Discover tab. If I don't have any interests it would show the "old" Discover content and only when moving back and forth between the tabs does it show the interests screen

Will be fixed in a subsequent task. :)

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift (hunk 16 above)
Comment (yaelirub) [reply WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:1907]: good call

File: WordPress/Classes/ViewRelated/Reader/Select Interests/ReaderSelectInterestsViewController.swift [hunk 20]
```diff
@@ -211,6 +225,38 @@ extension ReaderSelectInterestsViewController: UICollectionViewDelegateFlowLayou
 // MARK: - ReaderInterestsDataDelegate
 extension ReaderSelectInterestsViewController: ReaderInterestsDataDelegate {
     func readerInterestsDidUpdate(_ dataSource: ReaderInterestsDataSource) {
-        reloadData()
+        if dataSource.count > 0 {
+            hideLoadingView()
+            reloadData()
+        } else {
+            displayLoadingViewWithWebAction(title: "lalalalala")
```
Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/Select Interests/ReaderSelectInterestsViewController.swift:232]: lo what is this la la la la la ?

**Comment by yaelirub** [comment]
@leandroalonso ,
>Will be fixed in a subsequent task. :)

Great! please make sure that this doesn't slip off and there is an issue ticket for this.
I'm merging this one
